### PR TITLE
Feature/statistics

### PR DESCRIPTION
**Pull request by IILTCC**
finished creating statistics and sending them to kafka

### CODE REVIEW

File: telemetry_device/KafkaConnection.cs [hunk 1]
```diff
@@ -28,15 +30,23 @@ public KafkaConnection(TelemetryDeviceSettings telemetryDeviceSettings)
             _producer = new ProducerBuilder<Null, string>(producerConfig).Build();
             _adminClient = new AdminClientBuilder(adminConfig).Build();
         }
-        public void SendToTopic(string topicName, Dictionary<string,(int,bool)> paramDict)
+        public void SendIcdToTopic(string topicName, Dictionary<string,(int,bool)> paramDict)
```
Comment (eyal-otiker) [inline telemetry_device/KafkaConnection.cs:33]: Change name of func to SendFrameToTopic

File: telemetry_device/KafkaConnection.cs [hunk 2]
```diff
@@ -28,15 +30,23 @@ public KafkaConnection(TelemetryDeviceSettings telemetryDeviceSettings)
...
         {
-           
             string jsonString = JsonConvert.SerializeObject(paramDict);
             Message<Null,string> message = new Message<Null, string>
             {
                 Value = jsonString
             };
             _producer.Produce(topicName, message);
+        }
+        public void SendStatisticToTopic(Dictionary<MetricType,float> metricDict)
```
Comment (eyal-otiker) [inline telemetry_device/KafkaConnection.cs:42]: Change name of func to SendStatisticsToKafka

File: telemetry_device/KafkaConnection.cs
Comment (eyal-otiker) [inline telemetry_device/KafkaConnection.cs:1]: Unit the same code line in SendFrameToKafka and in SendStatisticsToKafka to one func

File: telemetry_device/compactCollection/MetricType.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace telemetry_device.compactCollection
+{
+    enum MetricType
```
Comment (eyal-otiker) [inline telemetry_device/compactCollection/MetricType.cs:9]: Change name of enum

File: telemetry_device/compactCollection/IcdtoMetricItem.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace telemetry_device.compactCollection
+{
+    class IcdtoMetricItem
```
Comment (eyal-otiker) [inline telemetry_device/compactCollection/IcdtoMetricItem.cs:9]: change name of class

File: telemetry_device/compactCollection/IcdtoMetricItem.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        public MetricType CorruptedPacket { get; set; }
+        public MetricType KafkaSend { get; set; }
```
Comment (eyal-otiker) [inline telemetry_device/compactCollection/IcdtoMetricItem.cs:12]: change name of parameter

File: telemetry_device/compactCollection/MetricType.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        SniffingTime,
+        PacketDropRate,
+        FlightBoxDownCorruptedPacket,
```
Comment (eyal-otiker) [inline telemetry_device/compactCollection/MetricType.cs:13]: This is not good that you are write for every icd type the static, find other way

File: telemetry_device/Pipeline.cs [hunk 7]
```diff
@@ -110,16 +137,33 @@ private TransformBlockItem ExtractPacketData(Packet packet)
             byte[] timestampBytes = new byte[24];
             for (int i = 0; i < timestampBytes.Length; i++)
                 timestampBytes[i] = udpPacket.PayloadData[i + 3];
+
             string timestamp = Encoding.ASCII.GetString(timestampBytes);
-            
+
+            // the format in which the timestamp is in the packet
+            DateTime dateTime = DateTime.ParseExact(timestamp, "dd,MM,yyyy,HH,mm,ss,ffff", CultureInfo.InvariantCulture);
```
Comment (eyal-otiker) [inline telemetry_device/Pipeline.cs:144]: put "dd,MM,yyyy,HH,mm,ss,ffff" as const

File: telemetry_device/Pipeline.cs [hunk 8]
```diff
@@ -32,17 +36,20 @@ class PipeLine
 
         private KafkaConnection _kafkaConnection;
         private TelemetryLogger _logger;
+        private StatisticsAnalyzer _statAnalyze;
+        private readonly Dictionary<IcdTypes,IcdtoStatItem> _icdToMetric;
```
Comment (eyal-otiker) [inline telemetry_device/Pipeline.cs:40]: change name of param

File: telemetry_device/KafkaConnection.cs [hunk 9]
```diff
@@ -28,15 +30,23 @@ public KafkaConnection(TelemetryDeviceSettings telemetryDeviceSettings)
             _producer = new ProducerBuilder<Null, string>(producerConfig).Build();
             _adminClient = new AdminClientBuilder(adminConfig).Build();
         }
-        public void SendToTopic(string topicName, Dictionary<string,(int,bool)> paramDict)
+        public void SendFrameToKafka(string topicName, Dictionary<string,(int,bool)> paramDict)
```
Comment (eyal-otiker) [inline telemetry_device/KafkaConnection.cs:33]: Unit the same code line in SendFrameToKafka and in SendStatisticsToKafka to one func

File: telemetry_device/Pipeline.cs [hunk 10]
```diff
@@ -119,7 +143,19 @@ private SendToKafkaItem ProccessPackets(TransformBlockItem transformItem)
...
+                DateTime beforeDecrypt = DateTime.Now;
+
                 Dictionary<string, (int, bool)> decryptedParamDict = _icdDictionary[transformItem.PacketType].DecryptPacket(transformItem.PacketData);
+                int decryptTime = (int)DateTime.Now.Subtract(beforeDecrypt).TotalMilliseconds;
+
+                _statAnalyze.UpdateStatistic(MultiStatisticType.DecryptTime, transformItem.PacketType, decryptTime);
+
+                int errorCounter = 0;
+                foreach ((int, bool) param in decryptedParamDict.Values)
+                    if (param.Item2)
```
Comment (eyal-otiker) [inline telemetry_device/Pipeline.cs:155]: Add comment what is Item2

File: telemetry_device/compactCollection/IcdtoStatItem.cs [hunk 11]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace telemetry_device.compactCollection
+{
+    class IcdtoStatItem
```
Comment (eyal-otiker) [inline telemetry_device/compactCollection/IcdtoStatItem.cs:9]: change name

File: telemetry_device/StatisticsAnalyzer.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+            Dictionary<string, float> avgDict = new Dictionary<string, float>();
+            foreach(SingleStatisticType key in _singleStatistics.Keys)
+            {
+                avgDict.Add(key.ToString(), _singleStatistics[key].GetAvg());
+            }
+            foreach(MultiStatisticType key in _multiStatistics.Keys)
+            {
+                foreach(IcdTypes icdType in Enum.GetValues(typeof(IcdTypes)))
+                {
+                    avgDict.Add(key.ToString()+"-"+icdType.ToString(),_multiStatistics[key].GetAvg(icdType));
```
Comment (eyal-otiker) [inline telemetry_device/StatisticsAnalyzer.cs:56]: why you put `-`?

File: telemetry_device/compactCollection/MultiStatisticType.cs [hunk 13]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace telemetry_device.compactCollection
+{
+    enum MultiStatisticType
```
Comment (eyal-otiker) [inline telemetry_device/compactCollection/MultiStatisticType.cs:9]: The word multi not very unsderstandable

File: telemetry_device/compactCollection/SingleStatisticType.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace telemetry_device.compactCollection
+{
+    enum SingleStatisticType
```
Comment (eyal-otiker) [inline telemetry_device/compactCollection/SingleStatisticType.cs:9]: The word single not very unsderstandable

File: telemetry_device/Pipeline.cs [hunk 15]
```diff
@@ -119,7 +143,19 @@ private SendToKafkaItem ProccessPackets(TransformBlockItem transformItem)
...
             try
             {
+                DateTime beforeDecrypt = DateTime.Now;
+
                 Dictionary<string, (int, bool)> decryptedParamDict = _icdDictionary[transformItem.PacketType].DecryptPacket(transformItem.PacketData);
+                int decryptTime = (int)DateTime.Now.Subtract(beforeDecrypt).TotalMilliseconds;
+
+                _statAnalyze.UpdateStatistic(MultiStatisticType.DecryptTime, transformItem.PacketType, decryptTime);
+
+                int errorCounter = 0;
```
Comment (eyal-otiker) [inline telemetry_device/Pipeline.cs:153]: put lines 153-156 in new func

File: telemetry_device/StatisticsAnalyzer.cs (hunk 12 above)
Comment (IILTCC) [reply telemetry_device/StatisticsAnalyzer.cs:56]: to make it more readable

Comment (eyal-otiker) [reply telemetry_device/StatisticsAnalyzer.cs:56]: It is not readable, use in object for the key

Comment (IILTCC) [reply telemetry_device/StatisticsAnalyzer.cs:56]: What do you mean use in object, creating my own class to act as the key?

Comment (eyal-otiker) [reply telemetry_device/StatisticsAnalyzer.cs:56]: yes

File: telemetry_device/compactCollection/MultiStatistic.cs [hunk 16]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using telemetry_device_main.icds;
+
+namespace telemetry_device.compactCollection
+{
+    class MultiStatistic
```
Comment (eyal-otiker) [inline telemetry_device/compactCollection/MultiStatistic.cs:10]: change name of the class

File: telemetry_device/StatisticsAnalyzer.cs [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            _globalStatistics[statType].AddCounter(value);
+        }
+        public void UpdateStatistic(IcdStatisticType statType,IcdTypes icdType,int value)
+        {
+            _icdStatistics[statType].AddCounter(icdType, value);
+        }
+        public Dictionary<string,float> GetDataDictionary()
+        {
+            Dictionary<string, float> avgDict = new Dictionary<string, float>();
```
Comment (eyal-otiker) [inline telemetry_device/StatisticsAnalyzer.cs:47]: Not good the key of the dictionary will be string, put the key as the type of the object

**Pull request merged**